### PR TITLE
AMDGPU/GlobalISel: Handle atomic sextload and zextload

### LINKED ISSUE #111645: AMDGPU: sextload/zextload atomics do not select in globalisel
```
; RUN: llc -global-isel -mtriple=amdgcn-amd-amdhsa -mcpu=gfx900 < %s

define i32 @sextload_atomic(ptr %p) {
  %ld = load atomic i8, ptr %p syncscope("agent-one-as") acquire, align 1
  %zext = sext i8 %ld to i32
  ret i32 %zext
}

define i32 @zextload_atomic(ptr %p) {
  %ld = load atomic i8, ptr %p syncscope("agent-one-as") acquire, align 1
  %zext = zext i8 %ld to i32
  ret i32 %zext
}
```

LLVM ERROR: cannot select: %11:vgpr_32(s32) = G_SEXTLOAD %0:vreg_64(p0) :: (load syncscope("agent-one-as") acquire (s8) from %ir.p) (in function: sextload_atomic)

### PR DESCRIPTION
**Pull request by arsenm**
Atomic loads are handled differently from the DAG, and have separate opcodes
and explicit control over the extensions, like ordinary loads. Add
new patterns for these.

There's room for cleanup and improvement. d16 cases aren't handled.

Fixes #111645

### CODE REVIEW

**Comment by arsenm** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/111721?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#111721** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/111721?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#111720** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/111720?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @arsenm and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: llvm/test/CodeGen/AMDGPU/GlobalISel/atomic_load_flat.ll [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+; GFX8-NEXT:    flat_load_sbyte v2, v[0:1] glc
+; GFX8-NEXT:    flat_load_ubyte v0, v[0:1] glc
+; GFX8-NEXT:    s_waitcnt vmcnt(0) lgkmcnt(0)
+; GFX8-NEXT:    v_mov_b32_e32 v0, v2
+; GFX8-NEXT:    s_setpc_b64 s[30:31]
+;
+; GFX9-LABEL: atomic_load_flat_monotonic_i8_sext_to_i32:
+; GFX9:       ; %bb.0:
+; GFX9-NEXT:    s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
+; GFX9-NEXT:    flat_load_sbyte v2, v[0:1] glc
```
Comment (rampitec) [inline llvm/test/CodeGen/AMDGPU/GlobalISel/atomic_load_flat.ll:51]: What happened here? Why is there 2 loads?

**Review by rampitec — COMMENTED** [review]
Missing test for buffer loads?

**Comment by arsenm** [comment]
> Missing test for buffer loads?

Those are the gfx7 global cases. There aren't any atomic buffer load intrinsics

File: llvm/test/CodeGen/AMDGPU/GlobalISel/atomic_load_flat.ll (hunk 1 above)
Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/GlobalISel/atomic_load_flat.ll:51]: There's a bug in amdgpu-postlegalizer-combiner, so it's not from the selection

**Comment by rampitec** [comment]
> > Missing test for buffer loads?
> 
> Those are the gfx7 global cases. There aren't any atomic buffer load intrinsics

But patch adds several MUBUF_Pseudo_Load_Pats which are not covered by tests?

**Comment by arsenm** [comment]
> But patch adds several MUBUF_Pseudo_Load_Pats which are not covered by tests?

The only cases that might have missing coverage is extend to 16-bit register cases. In the DAG we didn't have legal 16-bit types on gfx6/7, but we could handle the loads here

**Comment by arsenm** [comment]
ping

**Comment by rampitec** [comment]
> > But patch adds several MUBUF_Pseudo_Load_Pats which are not covered by tests?
> 
> The only cases that might have missing coverage is extend to 16-bit register cases. In the DAG we didn't have legal 16-bit types on gfx6/7, but we could handle the loads here

I still think these added MUBUF patterns need at least some test coverage.

**Comment by arsenm** [comment]
> I still think these added MUBUF patterns need at least some test coverage.

Just removed the 16-bit result ones

File: llvm/lib/Target/AMDGPU/BUFInstructions.td [hunk 2]
```diff
@@ -983,15 +983,20 @@ defm BUFFER_LOAD_LDS_U16 : MUBUF_Pseudo_Loads_LDSOpc <
 >;
 
 defm : MUBUF_Pseudo_Load_Pats<"BUFFER_LOAD_UBYTE", i32, atomic_load_8_global>;
+defm : MUBUF_Pseudo_Load_Pats<"BUFFER_LOAD_UBYTE", i32, atomic_load_zext_8_global>;
```
Comment (rampitec) [inline llvm/lib/Target/AMDGPU/BUFInstructions.td:986]: But these are still not covered?

**Review by arsenm — COMMENTED** [review]
ping

File: llvm/lib/Target/AMDGPU/BUFInstructions.td (hunk 2 above)
Comment (rampitec) [inline llvm/lib/Target/AMDGPU/BUFInstructions.td:986]: I still think if you are adding buffer patterns you need to have tests for them.

Comment (arsenm) [reply llvm/lib/Target/AMDGPU/BUFInstructions.td:986]: They are there

Comment (rampitec) [reply llvm/lib/Target/AMDGPU/BUFInstructions.td:986]: I see modified tests for flat, global and local.

Comment (arsenm) [reply llvm/lib/Target/AMDGPU/BUFInstructions.td:986]: This is not for buffer intrinsics or fat pointers. This is for global pointers implemented with mubuf instructions. This is covered by the global run lines for gfx6. The -flat-for-global case for gfx7 do not correctly use the mubuf results, but that's a separate issue

Comment (rampitec) [reply llvm/lib/Target/AMDGPU/BUFInstructions.td:986]: Ok

**Review by rampitec — APPROVED** [review]

**Comment by arsenm** [comment]
### Merge activity

* **Oct 31, 10:38 AM EDT**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/llvm/llvm-project/111721).

**Pull request merged**